### PR TITLE
move Deaggregate pass to post_borrowck_cleanup

### PR DESCRIPTION
**Pull request by RalfJung**
In https://github.com/rust-lang/rust/pull/70073 MIR pass handling got reorganized, but with the goal of not changing behavior (except for disabling some optimizations on opt-level = 0). But there we realized that the `Deaggregator` pass, while conceptually more of a "cleanup" pass (and one that should be run before optimizations), was run in the middle of the optimization chain. Likely this is an accident of history, so I suggest we try and clean that up by making it a proper cleanup pass.

This does change mir-opt output, likely because deaggregation now runs before const-prop instead of after. I am not sure if that is acceptable, desirable, or a problem.

r? @oli-obk @wesleywiser @eddyb @rust-lang/wg-mir-opt

### CODE REVIEW

**Comment by RalfJung** [comment]
I guess it is worth checking if this has any perf implications.
@bors rollup=never try
@rust-timer queue

File: src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff [hunk 1]
```diff
@@ -14,44 +14,35 @@
...
                                            // ty::Const
                                            // + ty: i32
                                            // + val: Value(Scalar(0x00000002))
                                            // mir::Constant
                                            // + span: $DIR/aggregate.rs:5:20: 5:21
                                            // + literal: Const { ty: i32, val: Value(Scalar(0x00000002)) }
--         _2 = (_3.1: i32);                // scope 0 at $DIR/aggregate.rs:5:13: 5:24
--         _1 = Add(move _2, const 0i32);   // scope 0 at $DIR/aggregate.rs:5:13: 5:28
-+         _2 = const 1i32;                 // scope 0 at $DIR/aggregate.rs:5:13: 5:24
+          _2 = (_3.1: i32);                // scope 0 at $DIR/aggregate.rs:5:13: 5:24
```
Comment (oli-obk) [inline src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: This is a regression in const prop, because `_3` is being assigned not in a single statement but in multiple, causing us to bail out. There's no fundamental reason to do so, but we need to come up with a new algorithm for figuring out which locals we can propagate. I think right now we bail out if a local has multiple assignments, but only the order of assignments is relevant I believe.

**Comment by RalfJung** [comment]
Curiously, [this causes incremental test suite failures](https://github.com/rust-lang/rust/runs/648966302) -- I am not sure what that is about, and have no idea how to debug it.

**Comment by oli-obk** [comment]
Hehe, I think that is because the incremental tests expect the changes done to not have an effect on MIR, but the "no effect" only existed because const prop eliminated all changes between the two variants. This is not true anymore due to the const prop regression.

**Comment by wesleywiser** [comment]
I recall removing annotations from the incremental tests indicating that some changes should have an effect on the optimized MIR when we switched const prop on by default. So that sounds correct to me.

**Comment by RalfJung** [comment]
That's good, it means we have just one side-effect of this change and not two. :D

So is this ConstProp effect something that we should fix before landing this PR? I'm afraid I won't have time to look into that, but I am happy to let this PR sit until someone gets around to that (or else turn it into an issue if we think that might take longer).

**Comment by oli-obk** [comment]
> So is this ConstProp effect something that we should fix before landing this PR?

I think we should fix it. This PR breaks all or at least most of our aggregate + const prop tests. Since we now have in-block const prop even for variables, we can actually write tests for this situation without your PR, so I'm going to do this change independently from your PR, get that merged, and then we merge your PR (whose diff should become much simpler after that).

**Comment by RalfJung** [comment]
@oli-obk all right, sounds good!

**Comment by oli-obk** [comment]
:tada: https://github.com/rust-lang/rust/pull/71953

**Comment by RalfJung** [comment]
Wow that's a *serious* perf regression for servo-opt, let's hope it gets fixed by your patch @oli-obk.

**Comment by oli-obk** [comment]
:( I'm just getting bad gateway errors when trying to open the perf result

**Comment by wesleywiser** [comment]
It's working now and *ouch!*

**Comment by oli-obk** [comment]
Looks like CGU problems *again* https://perf.rust-lang.org/detailed-query.html?commit=89530668eb34ae5856fd9aabf52a7c8afaa08808&base_commit=43271a39adc26d29350d4830b594c6435472815e&benchmark=script-servo-opt&run_name=patched%20incremental:%20println

https://github.com/rust-lang/compiler-team/issues/281 is really getting more and more important

**Comment by RalfJung** [comment]
I have no idea how to read that "detailed query" page, is there a guide for that somewhere? Somehow there's a "totals" row that sums up to either 136.02% of the time or 56.7% of something without a column header, where I'd expect a total to be 100%, so I am already lost in that row.

**Comment by RalfJung** [comment]
Oh wait maybe that weird `Δ` is not part of the GUI but the column header? I thought it indicates what things are sorted by (upwards arrow), but maybe it's the letter Delta?

**Comment by Mark-Simulacrum** [comment]
Yes, that's intended to be a delta -- I'm forgetting now why it's not just written out...

The total time header (e.g. 136.02% in this case) is the amount of time self profile claims it took divided by the amount of time that the cpu clock statistic gave us -- basically wall clock but accounting for parallelism. It can mostly be ignored, just gives you a rough sense how much time we're likely to be "off" by, though I can't tell you why we're sometimes off by so much.

**Comment by RalfJung** [comment]
Okay. But why are there three Delta columns that are very different? Is that delta in time, delta in "full executions" (are there "partial executions"?), and "loading delta" (whatever that is)?

**Comment by RalfJung** [comment]
Let's move this to a separate issue: https://github.com/rust-lang/rustc-perf/issues/648

**Comment by wesleywiser** [comment]
> rust-lang/compiler-team#281 is really getting more and more important

Adding this to my examples list.

**Comment by oli-obk** [comment]
The PR this was blocked on has been merged, please rebase and rerun perf

File: src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff [hunk 2]
```diff
@@ -14,44 +14,35 @@
...
+          (_3.2: i32) = const 2i32;        // scope 0 at $DIR/aggregate.rs:5:13: 5:22
                                            // ty::Const
                                            // + ty: i32
                                            // + val: Value(Scalar(0x00000002))
                                            // mir::Constant
                                            // + span: $DIR/aggregate.rs:5:20: 5:21
                                            // + literal: Const { ty: i32, val: Value(Scalar(0x00000002)) }
--         _2 = (_3.1: i32);                // scope 0 at $DIR/aggregate.rs:5:13: 5:24
--         _1 = Add(move _2, const 0i32);   // scope 0 at $DIR/aggregate.rs:5:13: 5:28
-+         _2 = const 1i32;                 // scope 0 at $DIR/aggregate.rs:5:13: 5:24
```
Comment (RalfJung) [inline src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: This seems to still regress -- `_2` does not get a constant value any more.

File: src/test/mir-opt/simplify-arm-identity/32bit/rustc.main.SimplifyArmIdentity.diff [hunk 3]
```diff
@@ -28,16 +28,34 @@
                                            // + literal: Const { ty: u8, val: Value(Scalar(0x00)) }
           discriminant(_1) = 0;            // scope 0 at $DIR/simplify-arm-identity.rs:18:18: 18:29
           StorageLive(_2);                 // scope 1 at $DIR/simplify-arm-identity.rs:19:18: 22:6
-          _3 = const 0isize;               // scope 1 at $DIR/simplify-arm-identity.rs:20:9: 20:20
+          _3 = discriminant(_1);           // scope 1 at $DIR/simplify-arm-identity.rs:20:9: 20:20
```
Comment (oli-obk) [inline src/test/mir-opt/simplify-arm-identity/32bit/rustc.main.SimplifyArmIdentity.diff:31]: cc @wesleywiser I think we need to adjust the simplify pass to handle deaggregated MIR

File: src/test/mir-opt/generator-tiny/rustc.main-{{closure}}.generator_resume.0.mir [hunk 4]
```diff
@@ -1,5 +1,5 @@
 // MIR for `main::{{closure}}#0` 0 generator_resume
-// generator_layout = GeneratorLayout { field_tys: [HasDrop], variant_fields: [[], [], [], [_0]], storage_conflicts: BitMatrix { num_rows: 1, num_columns: 1, words: [1], marker: PhantomData } }
```
Comment (oli-obk) [inline src/test/mir-opt/generator-tiny/rustc.main-{{closure}}.generator_resume.0.mir:2]: oh this is a nice side effect from running the deaggregator before `StateTransform`

Comment (RalfJung) [reply src/test/mir-opt/generator-tiny/rustc.main-{{closure}}.generator_resume.0.mir:2]: What is the effect? Less stuff in the generator?
Cc @jonas-schievink

Comment (jonas-schievink) [reply src/test/mir-opt/generator-tiny/rustc.main-{{closure}}.generator_resume.0.mir:2]: Looks like it the deaggregator "unpeels" empty types now, so this ZST doesn't need to be stored anymore (not that it matters too much for ZSTs).

Note that the generator transform uses the liveness dataflow analysis, which doesn't work correctly after the deaggregator has run - it treats every assignment to a field of an aggregate as a use of the base local, which may lead to locals being considered live up to the function's entry point. This does not cause unsoundness here, but it might cause some size regressions. However it looks like the `MaybeRequiresStorage` analysis fully mitigates this effect by only requiring storage for a local after the first field of the aggregate is written.

File: src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff [hunk 5]
```diff
@@ -14,44 +14,35 @@
...
+          _2 = (_3.1: i32);                // scope 0 at $DIR/aggregate.rs:5:13: 5:24
+          _1 = Add(move _2, const 0i32);   // scope 0 at $DIR/aggregate.rs:5:13: 5:28
```
Comment (RalfJung) [inline src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:39]: This still

File: src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff (hunk 2 above)
Comment (RalfJung) [inline src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: @oli-obk seems like even with your latest patch this still regresses?

Comment (oli-obk) [reply src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: ok, this is getting weird. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=387d515a8f94a2643d74169f9d80dfa5 is *basically* the same (except that there's an intial assignment to the entire local). Please try compiling this test with `RUSTC_LOG=rustc_mir::interpret::const_prop`, maybe that sheds some light on the problem

Comment (RalfJung) [reply src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: I think you mean `rustc_mir::transform::const_prop`:
```
[TRACE rustc_mir::transform::const_prop] ConstProp starting for DefId(0:3 ~ aggregate[317d]::main[0])
[TRACE rustc_mir::transform::const_prop] local _1 will only be propagated inside its block, because it's a user variable
[TRACE rustc_mir::transform::const_prop] local _3 can't be propagated because of multiple assignments
[TRACE rustc_mir::transform::const_prop] local _3 can't be propagated because of multiple assignments
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageLive(_1)
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageLive(_2)
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageLive(_3)
[TRACE rustc_mir::transform::const_prop] visit_statement: (_3.0: i32) = const 0i32
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const 0i32, place = (_3.0: i32))
[TRACE rustc_mir::transform::const_prop] can't propagate into (_3.0: i32)
[TRACE rustc_mir::transform::const_prop] visit_constant: const 0i32
[TRACE rustc_mir::transform::const_prop] visit_statement: (_3.1: i32) = const 1i32
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const 1i32, place = (_3.1: i32))
[TRACE rustc_mir::transform::const_prop] can't propagate into (_3.1: i32)
[TRACE rustc_mir::transform::const_prop] visit_constant: const 1i32
[TRACE rustc_mir::transform::const_prop] visit_statement: (_3.2: i32) = const 2i32
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const 2i32, place = (_3.2: i32))
[TRACE rustc_mir::transform::const_prop] can't propagate into (_3.2: i32)
[TRACE rustc_mir::transform::const_prop] visit_constant: const 2i32
[TRACE rustc_mir::transform::const_prop] visit_statement: _2 = (_3.1: i32)
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = (_3.1: i32), place = _2)
[TRACE rustc_mir::transform::const_prop] propagation into _2 failed.
                            Nuking the entire site from orbit, it's the only way to be sure
[TRACE rustc_mir::transform::const_prop] visit_statement: _1 = Add(move _2, const 0i32)
[TRACE rustc_mir::transform::const_prop] checking BinaryOp(op = Add, left = move _2, right = const 0i32)
[TRACE rustc_mir::transform::const_prop] propagation into _1 failed.
                            Nuking the entire site from orbit, it's the only way to be sure
[TRACE rustc_mir::transform::const_prop] visit_constant: const 0i32
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageDead(_2)
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageDead(_3)
[TRACE rustc_mir::transform::const_prop] visit_statement: _0 = const ()
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const (), place = _0)
[TRACE rustc_mir::transform::const_prop] replacing const () with OpTy { op: Immediate(Scalar(<ZST>)), layout: TyAndLayout { ty: (), layout: Layout { fields: Arbitrary { offsets: [], memory_index: [] }, variants: Single { index: 0 }, abi: Aggregate { sized: true }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 3 } }, size: Size { raw: 0 } } } }
[TRACE rustc_mir::transform::const_prop] skipping replace of Rvalue::Use(const () because it is already a const
[TRACE rustc_mir::transform::const_prop] propagated into _0
[TRACE rustc_mir::transform::const_prop] visit_constant: const ()
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageDead(_1)
[TRACE rustc_mir::transform::const_prop] ConstProp done for DefId(0:3 ~ aggregate[317d]::main[0])
```

Comment (oli-obk) [reply src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: ok thanks, this makes it clear what the problem is, I'll look into what the solution is.

Comment (oli-obk) [reply src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: I can't do this locally without your PR. There's no way to produce the relevant MIR without your PR :(

I believe it should get fixed by changing https://github.com/rust-lang/rust/blob/4b1f86adbe41e8dd4864ca2315f43953dd503bb5/src/librustc_mir/transform/const_prop.rs#L790-L803 to change `ConstPropMode` to `OnlyInsideOwnBlock` if it currently is `FullConstProp`.

Comment (RalfJung) [reply src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: Feel free to push to my branch -- we can make this a collaborative PR, maybe @jonas-schievink can review then.

Comment (oli-obk) [reply src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff:38]: you need to check the box on the right allowing edits or give me access to your repo in order for me to push

**Comment by RalfJung** [comment]
Weird, I thought that is checked by default... should be checked now.

**Comment by oli-obk** [comment]
still can't push. I now pushed my commit to https://github.com/oli-obk/rust/tree/deaggregate-is-cleanup

**Comment by RalfJung** [comment]
Weird, the checkbox was unchecked again. Looks like it did not register... now it seems to stick.

**Comment by RalfJung** [comment]
Nope. It goes away each time I reload. Wtf.

**Comment by RalfJung** [comment]
I now manually added your commit here.

**Comment by RalfJung** [comment]
That is curious:
```
2020-05-31T19:57:07.7282130Z error: internal compiler error: error during interning should later cause validation failure
2020-05-31T19:57:07.7282669Z   --> /checkout/src/test/mir-opt/box_expr.rs:19:9
2020-05-31T19:57:07.7282887Z    |
2020-05-31T19:57:07.7283054Z 19 |         println!("splat!");
2020-05-31T19:57:07.7283256Z    |         ^^^^^^^^^^^^^^^^^^^
2020-05-31T19:57:07.7283415Z    |
2020-05-31T19:57:07.7284009Z    = note: this error: internal compiler error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
2020-05-31T19:57:07.7296760Z 
2020-05-31T19:57:07.7297773Z thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:366:17
```

**Comment by RalfJung** [comment]
I rebased on top of current master.

**Comment by oli-obk** [comment]
I have a fix, you'll need to pull it from my branch again

**Comment by RalfJung** [comment]
Done that.

**Comment by oli-obk** [comment]
Heads up @rust-lang/wg-mir-opt I changed const-prop to stop interning constants, as it was trying to do so with partially initialized aggregates. The change is https://github.com/oli-obk/rust/commit/d18a704cc55686711d202e6508414cacf7c41cc8 . Doing so had no effect on our mir-opt test suite.

**Comment by RalfJung** [comment]
Looks like mir-opt passes now. :)

But there is a codegen test failure:
```
2020-06-03T15:00:14.6144951Z ---- [codegen] codegen/consts.rs stdout ----
2020-06-03T15:00:14.6145098Z 
2020-06-03T15:00:14.6145408Z error: verification with 'FileCheck' failed
2020-06-03T15:00:14.6147079Z status: exit code: 1
2020-06-03T15:00:14.6147912Z command: "/usr/lib/llvm-8/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll" "/checkout/src/test/codegen/consts.rs"
2020-06-03T15:00:14.6148325Z stdout:
2020-06-03T15:00:14.6148750Z ------------------------------------------
2020-06-03T15:00:14.6148920Z 
2020-06-03T15:00:14.6149285Z ------------------------------------------
2020-06-03T15:00:14.6149656Z stderr:
2020-06-03T15:00:14.6149996Z ------------------------------------------
2020-06-03T15:00:14.6150324Z /checkout/src/test/codegen/consts.rs:13:11: error: CHECK: expected string not found in input
2020-06-03T15:00:14.6150790Z // CHECK:***@alloc7 = {{.*}}, align 2
2020-06-03T15:00:14.6151141Z           ^
2020-06-03T15:00:14.6152140Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:11:1: note: scanning from here
2020-06-03T15:00:14.6152513Z @alloc8 = private unnamed_addr constant <{ [4 x i8] }> zeroinitializer, align 2
2020-06-03T15:00:14.6152708Z ^
2020-06-03T15:00:14.6153356Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:11:64: note: possible intended match here
2020-06-03T15:00:14.6153755Z @alloc8 = private unnamed_addr constant <{ [4 x i8] }> zeroinitializer, align 2
2020-06-03T15:00:14.6154043Z                                                                ^
2020-06-03T15:00:14.6154739Z /checkout/src/test/codegen/consts.rs:46:12: error: CHECK: expected string not found in input
2020-06-03T15:00:14.6155656Z  // CHECK: load %"E<i16, [i16; 3]>"*, %"E<i16, [i16; 3]>"** bitcast (<{ i8*, [0 x i8] }>* [[LOW_HIGH]] to %"E<i16, [i16; 3]>"**),
2020-06-03T15:00:14.6156046Z            ^
2020-06-03T15:00:14.6157095Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:41:28: note: scanning from here
2020-06-03T15:00:14.6157407Z define i64 @low_align_const() unnamed_addr #0 {
2020-06-03T15:00:14.6157604Z                            ^
2020-06-03T15:00:14.6158175Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:41:28: note: uses undefined variable "LOW_HIGH"
2020-06-03T15:00:14.6158536Z define i64 @low_align_const() unnamed_addr #0 {
2020-06-03T15:00:14.6158748Z                            ^
2020-06-03T15:00:14.6159427Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:44:2: note: possible intended match here
2020-06-03T15:00:14.6159922Z  %_2 = load %"E<i16, [i16; 3]>"*, %"E<i16, [i16; 3]>"** bitcast (<{ i8*, [0 x i8] }>* @1 to %"E<i16, [i16; 3]>"**), align 8, !nonnull !2
2020-06-03T15:00:14.6160227Z  ^
2020-06-03T15:00:14.6160500Z /checkout/src/test/codegen/consts.rs:54:12: error: CHECK: expected string not found in input
2020-06-03T15:00:14.6160981Z  // CHECK: load %"E<i16, i32>"*, %"E<i16, i32>"** bitcast (<{ i8*, [0 x i8] }>* [[LOW_HIGH]] to %"E<i16, i32>"**),
2020-06-03T15:00:14.6161321Z            ^
2020-06-03T15:00:14.6161858Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:54:29: note: scanning from here
2020-06-03T15:00:14.6162508Z define i64 @high_align_const() unnamed_addr #0 {
2020-06-03T15:00:14.6162705Z                             ^
2020-06-03T15:00:14.6163433Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:54:29: note: uses undefined variable "LOW_HIGH"
2020-06-03T15:00:14.6163772Z define i64 @high_align_const() unnamed_addr #0 {
2020-06-03T15:00:14.6163973Z                             ^
2020-06-03T15:00:14.6164503Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/consts/consts.ll:57:6: note: possible intended match here
2020-06-03T15:00:14.6165077Z  %_2 = load %"E<i16, i32>"*, %"E<i16, i32>"** bitcast (<{ i8*, [0 x i8] }>* @1 to %"E<i16, i32>"**), align 8, !nonnull !2
```

**Comment by oli-obk** [comment]
I pushed a fix to my repo's branch

**Comment by RalfJung** [comment]
Now `incremental/hashes/while_let_loops.rs` fails...^^

**Comment by oli-obk** [comment]
I adjusted the incremental test in my branch

**Comment by jonas-schievink** [comment]
Let me know if you need a review of this

**Comment by oli-obk** [comment]
@jonas-schievink I believe this is ready for review now

**Comment by RalfJung** [comment]
@oli-obk btw, do you know why the incremental tests needed these adjustments?

File: src/librustc_mir/transform/const_prop.rs [hunk 6]
```diff
@@ -707,11 +707,7 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                 ScalarMaybeUninit::Scalar(l),
                 ScalarMaybeUninit::Scalar(r),
             )) => l.is_bits() && r.is_bits(),
-            interpret::Operand::Indirect(_) if mir_opt_level >= 2 => {
-                let mplace = op.assert_mem_place(&self.ecx);
-                intern_const_alloc_recursive(&mut self.ecx, InternKind::ConstProp, mplace, false);
```
Comment (RalfJung) [inline src/librustc_mir/transform/const_prop.rs:712]: Why was this needed / why is this no longer needed? Seems like a surprising change to me.

Comment (oli-obk) [reply src/librustc_mir/transform/const_prop.rs:712]: This was buggy code that basically wasn't doing anything before this PR and was causing crashes with this PR. It's a leftover that multiple refactorings made obsolete. I can remove it on master if you prefer.

Comment (RalfJung) [reply src/librustc_mir/transform/const_prop.rs:712]: No it's fine. One less interning entry point to worry about. :)

**Comment by oli-obk** [comment]
> @oli-obk btw, do you know why the incremental tests needed these adjustments?

yes, they were unnecessary and fallout from deaggregating `_x = None` to `discriminant(_x) = 0`. I fixed this by const propping discriminant assignments. Please fetch the latest commits from my branch

**Comment by RalfJung** [comment]
Done. I also rebased to squash remove the reverted commit.

File: src/librustc_mir/interpret/step.rs [hunk 7]
```diff
@@ -74,7 +74,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         Ok(true)
     }
 
-    fn statement(&mut self, stmt: &mir::Statement<'tcx>) -> InterpResult<'tcx> {
+    crate fn statement(&mut self, stmt: &mir::Statement<'tcx>) -> InterpResult<'tcx> {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/step.rs:77]: Hm... of it's (semi-)public, maybe it should have a doc comment?

Also for consistency, `terminator` should probably get the same treatment.

**Comment by oli-obk** [comment]
ok... this is causing stage 2 failures now... Idk how that change could be misoptimizing anything and I won't have time to dig into it until some time next week.

**Comment by RalfJung** [comment]
Maybe it is indeed better to make that a separate PR, then.

(Quite amazing how hard it can be to reorder these passes.^^)

**Comment by oli-obk** [comment]
The change would likely not be misoptimizing anything without moving the deaggregation pass as early as you did

**Comment by oli-obk** [comment]
but we can move the independent things out of this PR, sure

**Comment by RalfJung** [comment]
Whatever you prefer. If you want you can also just make a PR yourself with this branch, at this point you did much more of the work than I did and this is way out of my league anyway.^^

**Comment by wesleywiser** [comment]
If you don't mind, I'd be happy to pull the "remove const prop interning" change into a separate PR. I suspect we can close a number of ICE issues with that change.

**Comment by oli-obk** [comment]
That would be great! Thanks

**Comment by RalfJung** [comment]
Does https://github.com/rust-lang/rust/pull/71911#discussion_r436273388 mean we should stop attempting this reordering?

**Comment by oli-obk** [comment]
I opened https://github.com/rust-lang/rust/issues/73203

I don't think we should stop this PR, the discussion is a separate one that we'll be able to resolve in either direction even with this change.

**Comment by RalfJung** [comment]
Status: blocked on https://github.com/rust-lang/rust/pull/73130.

**Comment by RalfJung** [comment]
#73130 landed. @oli-obk I don't know how many of your patches that included, which commits should be rebased?

**Comment by oli-obk** [comment]
I can do the rebase. Since I still can't modify this PR I'll do what you suggested earlier

>  If you want you can also just make a PR yourself with this branch

**Comment by RalfJung** [comment]
Yeah I'm happy for you to take over. :)
Closing this PR then.
